apps/spruce/src/pages/waterfall/caching/index.ts

```diff
@@ -0,0 +1,146 @@
+import { FieldMergeFunction, FieldReadFunction } from "@apollo/client";
+import { WaterfallQuery } from "gql/generated/types";
+import { VERSION_LIMIT } from "../constants";
+
+export const readVersions = ((existing, { args, readField }) => {
+  if (!existing) {
+    return undefined;
+  }
+
+  const minOrder = args?.options?.minOrder ?? 0;
+  const maxOrder = args?.options?.maxOrder ?? 0;
+  const limit = args?.options?.limit ?? VERSION_LIMIT;
+  const { mostRecentVersionOrder } =
+    readField<WaterfallQuery["waterfall"]["pagination"]>(
+      "pagination",
+      existing,
+    ) ?? {};
+
+  const existingVersions =
+    readField<WaterfallQuery["waterfall"]["flattenedVersions"]>(
+      "flattenedVersions",
+      existing,
+    ) ?? [];
+
+  const idx = existingVersions.findIndex((v) => {
+    const versionOrder = readField<number>("order", v) ?? 0;
+    if (minOrder) {
+      return versionOrder - 1 === minOrder;
+    }
+    if (maxOrder) {
+      return versionOrder + 1 === maxOrder;
+    }
+    return false;
+  });
+
+  if (idx === -1) {
+    return undefined;
+  }
+
+  let startIndex = maxOrder ? idx : 0;
+  let endIndex = maxOrder ? existingVersions.length : idx;
+  let numActivated = 0;
+
+  // Count backwards for paginating backwards.
+  if (minOrder) {
+    for (let i = endIndex; i >= 0; i--) {
+      if (readField<boolean>("activated", existingVersions[i])) {
+        numActivated += 1;
+        if (numActivated === limit) {
+          startIndex = i;
+          if (i > 0) {
+            // Check for leading inactive versions
+            if (
+              [...Array(i).keys()].every(
+                (index) =>
+                  readField<boolean>("activated", existingVersions[index]) ===
+                  false,
+              )
+            ) {
+              startIndex = 0;
+            }
+          }
+          break;
+        }
+      }
+    }
+    if (numActivated < limit) {
+      return undefined;
+    }
+  }
+
+  // Count forwards for paginating forwards.
+  if (maxOrder) {
+    for (let i = startIndex; i < existingVersions.length; i++) {
+      if (readField<boolean>("activated", existingVersions[i])) {
+        numActivated += 1;
+        if (numActivated === limit) {
+          endIndex = i;
+          break;
+        }
+      }
+    }
+    if (numActivated < limit) {
+      return undefined;
+    }
+  }
+
+  // Add 1 because slice is [inclusive, exclusive).
+  const flattenedVersions = existingVersions.slice(startIndex, endIndex + 1);
+
+  const zerothOrder = readField<number>("order", flattenedVersions[0]) ?? 0;
+  const prevOrderNumber =
+    mostRecentVersionOrder === zerothOrder
+      ? 0
+      : (readField<number>("order", flattenedVersions[0]) ?? 0);
+  const lastVersionOrder = readField<number>(
+    "order",
+    flattenedVersions[flattenedVersions.length - 1],
+  );
+  const nextOrderNumber =
+    !lastVersionOrder || lastVersionOrder === 1 ? 0 : lastVersionOrder;
```

Comment:
i think you can do a small simplification here:
```suggestion
  const lastVersionOrder =
    readField<number>(
      "order",
      flattenedVersions[flattenedVersions.length - 1],
    ) ?? 0;
  const nextOrderNumber = lastVersionOrder === 1 ? 0 : lastVersionOrder;
```